core/src/dsnp/graph_page.rs

```diff
@@ -0,0 +1,215 @@
+#![allow(dead_code)] // todo: remove
+use crate::dsnp::{api_types::*, dsnp_types::*};
+use std::{
+	collections::HashMap,
+	time::{SystemTime, UNIX_EPOCH},
+};
+
+/// Graph page structure
+#[derive(Debug, Clone, PartialEq)]
+pub struct GraphPage {
+	/// List of PRIds
+	prids: Vec<DsnpPrid>,
+	/// List of connections
+	connections: Vec<DsnpGraphEdge>,
+}
+
+/// Graph structure to hold pages of connections of a single type
+#[derive(Debug, Clone)]
+pub struct Graph {
+	pages: HashMap<PageId, GraphPage>,
+}
+
+/// Structure to hold all of a User's Graphs, mapped by ConnectionType
+#[derive(Debug, Clone)]
+pub struct UserGraph {
+	graphs: HashMap<ConnectionType, Graph>,
+}
+
+impl UserGraph {
+	/// Get the user's graph for the specified ConnectionType
+	fn get_graph(&self, connection_type: &ConnectionType) -> Option<&Graph> {
+		self.graphs.get(connection_type)
+	}
+}
+
+impl GraphPage {
+	/// Create a new, empty page
+	pub fn new() -> Self {
+		Self { prids: Vec::<DsnpPrid>::new(), connections: Vec::<DsnpGraphEdge>::new() }
+	}
+
+	/// Getter for the prids in the page
+	pub fn prids(&self) -> &Vec<DsnpPrid> {
+		&self.prids
+	}
+
+	/// Getter for the connections in the page
+	pub fn connections(&self) -> &Vec<DsnpGraphEdge> {
+		&self.connections
+	}
+
+	/// Setter for the prids in the page
+	pub fn set_prids(&mut self, prids: Vec<DsnpPrid>) {
```

Comment:
I think we would also need add_prid for a usecase that we are adding a new connection for private graph.

Do you think maybe we need different traits for different kind of graphs? Like prid only makes sense for Private and Friendship graphs. If we can make it happen easily I'll prefer it but if it's a significant change I think we can ignore it.

Reply:
Hmm, I was thinking that prids would only be set _en masse_, either on import or a complete page refresh of prids, as we were discussing. We already have a `set_prids` method, I guess we just need an `update_prids` method, that would iterate through the connections, calculate all the prids, and set into the page. I think maybe that's the subject of another PR.

Reply:
So we are not going to calculate a separate PRid for that connection. We would need to recalculate for all others to make sure that we are not going to have stale ones.